Use pytest for more TestSuite tests

orlitzky opened the pull request:
Fix some "slow doctest" warning by moving these out of the doctests, and into pytest where (IMO) they belong. They aren't for end-user consumption, and we aren't checking the output from them (as is the point of a doctest).

orlitzky committed:
src.sage.rings.valuation.mapped_valuation: drop two TestSuite() tests

These have been moved to pytest and no longer need to be run as
doctests.

orlitzky committed:
sage.rings.valuation.mapped_valuation_test: add pytests

Copy two TestSuite() runs from a TESTS:: block to pytest. These are on
the verge of hitting the "slow doctest" limit for long tests, and
sometimes cause warnings in the CI. But more importantly, they are
simply better suited to pytest: the user would never want to run them,
and the output is not meaningful as documentation.

orlitzky committed:
sage.rings.padics.padic_lattice_element_test: new pytests

Move some of the TestSuite() tests from a TESTS block to pytest.
These are slow, and not really meant for end-user consumption,
so are more appropriate here.

orlitzky committed:
sage.rings.padics.padic_lattice_element: delete old TESTS

Delete a few doctests that have been moved to pytest.

orlitzky committed:
sage.combinat.positive_integer_semigroup_test: new pytests

Move some TestSuite tests from src/sage/combinat/backtrack.py (for the
PositiveIntegerSemigroup class) to a new pytest file. These tests
aren't meant for end-user consumption, and they are slow, so they
are more appropriate here.

orlitzky committed:
sage.combinat.backtrack: delete old doctest

Remove a TestSuite() doctest that has been moved to pytest.

orlitzky committed:
sage.rings.padics.padic_lattice_element: add back FutureWarnings

Shuffling around the doctests has made some FutureWarnings happen in
different places.

orlitzky committed:
src/sage/rings/padics/meson.build: add padic_lattice_element_test.py

Run tools/update-meson.py to install the new pytest source file.

orlitzky committed:
src/sage/combinat/meson.build: add positive_integer_semigroup_test.py

Run tools/update-meson.py to install the new pytest source file.

tobiasdiez commented on src/sage/rings/padics/padic_lattice_element_test.py:
```diff
@@ -0,0 +1,27 @@
+        R3 = QpLC(2)
+        R4 = QpLF(2)
+
+    for R in (R1, R2, R3, R4):
```
Use paramatrized tests instead of the internal for loop?

orlitzky committed:
sage.rings.padics.padic_lattice_element_test: parameterize tests

Use parameterized tests and fixtures instead of an internal "for" loop
to improve the reporting a bit.

orlitzky committed:
sage.rings.valuation.mapped_valuation_test: parameterize tests

Use parameterized tests and fixtures to improve the reporting from
these tests.

orlitzky committed:
sage.misc.dev_tools.find_objects_from_name: use a copy of sys.modules

In an attempt to avoid the following error on the CI...

  dt.find_objects_from_name('FareySymbol')
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sage/src/sage/misc/dev_tools.py", line 279,
  in find_objects_from_name
    for smodule_name, smodule in sys.modules.items():
  RuntimeError: dictionary changed size during iteration

we make a copy of the sys.modules dict before iterating over its
items.

tobiasdiez commented on src/sage/combinat/positive_integer_semigroup_test.py:
```diff
@@ -0,0 +1,16 @@
+    PP = PositiveIntegerSemigroup()
+
+    # fewer max_runs since these are kind of slow
+    TestSuite(PP).run(verbose=True,
```
Is the verbose = True option needed for the functionality? If not, I think, I would prefer the quiet version, or maybe couple it to pytests verbosity level via `pytestconfig.get_verbosity() > 0 (or 1?)`

orlitzky replied:
Pytest already swallows stdout unless you pass `-s`, so i was thinking the added verbosity would only matter if the tests fail. I guess we could go back and make them all depend on the pytest verbosity though. I don't feel strongly about it either way.

tobiasdiez replied:
I noticed it mostly in CI, where it bloats the log without adding too much value (at least when everything passes). For the moment, that's okay but might get more annoying once more tests are migrated.

So maybe I do this later if I get annoyed enough - or I'll actually extract all of these testsuite _test methods (in the actual classes) to pytest.

orlitzky replied:
We could probably get away without `-s` in the CI. I think pytest will still dump the output that it captured in the event of a failure, while leaving everything nice and quiet if it passes.

tobiasdiez commented on src/sage/rings/padics/padic_lattice_element_test.py:
```diff
@@ -0,0 +1,65 @@
+
+@pytest.fixture
+def R1():
+    from warnings import catch_warnings, filterwarnings
```
What's the reason for the local imports?

orlitzky replied:
Though it's ugly, my reasoning was that the fixtures are loaded on-demand by the tests, and in the future we may run only a subset of tests (requiring a subset of fixtures). In that case the local imports won't slow things down.

tobiasdiez replied:
Are these imports really so expensive that it's worth optimizing this (at the cost of readability)?

orlitzky replied:
:shrug: I moved them to the top level

tobiasdiez reviewed (commented):
LGTM. I still have two questions, but they are not major and don't need to be 'resolved' before merging.

orlitzky committed:
sage.rings.padics.padic_lattice_element_test: improve readability

Move the padic imports and FutureWarning filter to the top level,
so we don't have to repeat them four times.

tobiasdiez reviewed (approved):
okay, let's get it in

user202729 commented on src/sage/rings/padics/padic_lattice_element_test.py:
```diff
@@ -0,0 +1,54 @@
+elements = ( "R1", "R2", "R3", "R4" )
+
+
+@pytest.mark.parametrize("e", elements)
```
the amount of boilerplate is scary.

Why can't you parametrize e over [ZpLC(2), ZpLF(2), etc.] directly?

orlitzky replied:
There are a couple of related problems. (I'm not happy about the boilerplate either, but it should decrease in relative volume as more tests are moved to pytest.)

The big one is that we want to cache the fixtures and create them on-demand, so they aren't recreated multiple times, or created when they won't be used (if only a subset of tests is being run). For that magic to work, we need to mark the fixtures with `@pytest.fixture`... but then you can't call them directly:

```
_____ ERROR collecting src/sage/rings/padics/padic_lattice_element_test.py _____
Fixture "R1" called directly. Fixtures are not meant to be called directly,
but are created automatically when test functions request them as parameters.
```

If we forego the caching / on-demand creation, we can create the objects ourselves, but then we lose the nice fixture naming. if I pass in `ZpLC(2)`, for example, we get...

```
src/sage/rings/padics/padic_lattice_element_test.py::test_padic_lattice_element[e0]...
```

instead of

```
src/sage/rings/padics/padic_lattice_element_test.py::test_padic_lattice_element[R1]...

```

which uses the fixture name. Now obviously `R1` is a lame name for a fixture and doesn't improve anything, but I'm trying to structure these in a decent way for future copy/pasting where the fixtures may actually be meaningful and take a while to construct.

user202729 replied:
"so that they aren't recreated multiple times"? Aren't they only used in this one test anyway?

tobiasdiez replied:
The idea of fixtures is that they can be easily reused in other tests (either in the same file or other test files), and pytest then makes sure that only one instance is created for all those tests. This obviously only scales once more and more tests are migrated to pytest. (eg here if all the `_test_xyz` methods from the actual class are moved into the pytest file. 

See eg symplectic forms tests for a more complicated example that would be hard to realize without fixtures.

orlitzky replied:
Right, there's only one test using it because there's only one test. But there are other doctests that construct (say) `ZpLC(2)`, and we could re-use the existing object in that case. Unlike with doctests, we don't have to reset the namespace before every test, so there's lots of room to reduce duplication.

If it was trivial to parameterize a test with fixtures, I probably wouldn't have bothered, since n >= 2 is usually where you'd stop and refactor. But it took me a few hours to figure out the trick, so I figure the polite thing to do is to set everything up nicely for the next person who wants to add a test using one of these lattice elements.

user202729 replied:
doctest doesn't need to reset the namespace before every test either. But usually reusing variables from another doctest would be confusing.

If you do say

```
sage: R1 = ZpLC(2)
sage: R1.test1()

another test::

sage: R1.test2()
```

then you're no worse off than the situation with pytest (and it's shorter…)

> pytest then makes sure that only one instance is created for all those tests

aren't the different files tested in parallel?

tobiasdiez replied:
> doctest doesn't need to reset the namespace before every test either. But usually reusing variables from another doctest would be confusing.
> 
> If you do say
> 
> ```
> sage: R1 = ZpLC(2)
> sage: R1.test1()
> 
> another test::
> 
> sage: R1.test2()
> ```
> 
> then you're no worse off than the situation with pytest (and it's shorter…)

This only works if those doctests are for the same method. If you want to test different methods, you would need to create a new instance of `ZpLC` in each of them. And this happens actually quite a lot: `ZpLC(2)` is actually created 35 times during the doctests, with pytest it would be only once (I know some instances of `ZpLC(2)` are actually in EXAMPLES, and you would like to keep them; but that only changes the numbers not the point of the argument)

> > pytest then makes sure that only one instance is created for all those tests
> 
> aren't the different files tested in parallel?

Not sure about the details, but it automatically groups tests that make use of the same fixture.

user202729 commented on src/sage/rings/valuation/mapped_valuation_test.py:
```diff
@@ -0,0 +1,55 @@
+extensions = ( "w0", "w1" )
+
+
+@pytest.mark.parametrize("e", extensions)
```
This part can do parametrize("index", [0, 1]) I guess.

orlitzky replied:
Ok, this one was easy to improve without hurting anything. I rebased onto the latest develop, but only the final commit is new.

user202729 commented on src/sage/rings/valuation/mapped_valuation_test.py:
```diff
@@ -0,0 +1,55 @@
+def test_finite_extension_from_limit_valuation(e, request):
+    r"""
+    Run the ``TestSuite()`` for two examples given in the
+    ``FiniteExtensionFromLimitValuation`` documentation.
```
Speaking of more downsides of pytest migration, a cost of having the test far from the code is that now if the example given in the FiniteExtensionFromLimitValuation documentation is changed, this docstring will silently become false.

You can change the docstring to "run the test suite for two examples that were given in the FiniteExtensionFromLimitValuation documentation", but then what's the value of the docstring?

orlitzky replied:
This is just laziness on my part and not inherent to pytest. I'd love to describe the objects being tested more intelligently, or explain why they're being tested... but the original doctests didn't come with an explanation, and I don't know anything about mapped valuations, so this was all I could think of.

orlitzky committed:
sage.rings.valuation.mapped_valuation_test: eliminate boilerplate

The one test in this file is easy to parameterize based on the index
(an int) of a fixture supplied to it. This allows us to eliminate a
lot of the boilerplate associated with fixture parameterizations.
